packages/block-editor/src/components/block-switcher/test/index.js

```diff
@@ -220,11 +240,8 @@ describe( 'BlockSwitcherDropdownMenu', () => {
 				} )
 			).toBeVisible();
 
-			const menu = screen.getByRole( 'menu', {
-				name: 'Block Name',
-			} );
 			expect( menu ).toBeInTheDocument();
-			expect( menu ).not.toBeVisible();
+			expect( menu ).toBeVisible();
```

Comment:
At this moment, if you check it with a snapshot, the opacity is not yet 1, but "only" 0.9994. There is an animation frame loop (in `framer-motion`) that gradually changes it from 0 to 1. At the moment of this check, i.e., after we're finished waiting for popover positioning, the animation hasn't finished yet.

We should do another wait with:
```js
await waitFor( () => expect(popover).toHaveStyle( { opacity: 1 } );
```
Or integrate it into `toBePositionedPopover`? Making it do a `toBePositionedAndFullDisplayedPopover`?

Reply:
Yep, I was thinking about that. I think this is a worthy improvement, let me work on it.

Reply:
Okay, this is interesting. I haven't debugged thoroughly yet, but the moment I add the assertion you mentioned, or extend `.toBePositionedPopover()` to check for `opacity: 1`, I get the `act` warnings again!

Reply:
I debugger this and the `act()` warning is caused by this `setHasAnimatedOnce` state update:

https://github.com/WordPress/gutenberg/blob/5fcf00da5ab1bfb64e1bbdd066f061c7daff7a7a/packages/components/src/popover/index.tsx#L120-L123

When `framer-motion` is finishing an animation, it will set the final `opacity: 1` style, then will resolve a promise, and then there's a `.then( () => props.onAnimationComplete() )` handler. The state update runs one promise tick after the `waitFor` for `opacity: 1` has finished.

@ciampo I'm curious what are we preventing with this `hasAnimatedOnce` guard? My understanding is that with a `<motion.div animate={ { opacity: 1 } }>`, the `animate` animation runs only once, on mount, and then never again. But a code comment suggests that it runs also on "subsequent prop changes"? What kind of change exactly can re-trigger the `animate` animation to run again?

If `Popover` didn't have the `hasAnimatedOnce` state, waiting for `.toHaveStyle( { opacity: 1 } )` would work flawlessly.

Reply:
Hey @jsnajdr , if I recall correctly that state variable is in place to avoid the animation from re-running when the `placement` changes, as this inline comment says:

https://github.com/WordPress/gutenberg/blob/b1202138f4e44fa6312e93f2e7d5132cf73586cd/packages/components/src/popover/index.tsx#L109-L112

If I remember correctly, without this flag, the animation would run again if the value of the `placement` prop changes.

Feel free to play around with it and remove it if we find a better solution to the problem!

Reply:
> without this flag, the animation would run again if the value of the `placement` prop (and probably other props too) change.

This is however not true, at least with the version of `framer-motion` that we're using now. The `animate` animation runs only on mount, never anywhere else, like on a prop change. When `placement` changes, then via `placementToMotionAnimationProps` the value of the `animate` prop changes, too, but that doesn't restart the animation either.

I verified that when debugging `framer-motion` back in December, and also experimentally. On this video, the block inserter popover appears with animation (I made it run slower to make it more visible), but then the switches between `top` and `bottom` placements are not animated any more:

![Screen Capture on 2023-01-03 at 14-05-27](https://user-images.githubusercontent.com/664258/210380864-f0e5ce51-3bd3-4890-a691-6ba0057cd55b.gif)

I think we can either remove `hasAnimatedOnce` completely, or use it just to set the `is-positioned` class (meaning that the popover has finished positioning and animating).

Reply:
Sounds good to me, then! Please do ping me on the PR if you're going to work on it, happy to help with reviewing and testing.